# Выплаты по кредиту. Решение трансцендентного уравнения

## Условие

Предположим, что кто-то желает взять кредит на сумму $\$100000$.
Он выбирает среди следующих возможностей:

 1) взять кредит на $10$ лет с ежемесячными выплатами в $\$1500$;
 
 2) взять кредит на $15$ лет с ежемесячными выплатами в $\$1250$;
 
 3) взять кредит на $20$ лет с ежемесячными выплатами в $\$1160$.

Какой кредит выгоднее брать?

Можно просто вычислить переплату, а можно еще учесть процентные ставки.

Сами процентные ставки здесь не указаны, но можно их найти из *основного финансового уравнения*:

$$
{\rm PV}\, (1+i)^n + {\rm PMT}\, (1+i)^{n-1} + {\rm PMT}\, (1+i)^{n-2} + \dots + {\rm PMT} + {\rm FV} = 0, \tag{1}
$$
связывающего следующие величины:
 - ${\rm PV}$ - величина кредита (present value);
 - ${\rm FV}$ - баланс после $n$ платежей (future value);
 - ${\rm PMT}$ - величина периодического платежа со знаком минус (payment);
 - $n$ - количество платежей;
 - $i$ - процентная ставка (interest rate).
 
Используя формулу для суммы элементов геометрической прогрессии, уравнение (1) запишем в следующем виде:
$$
{\rm PV}\, (1+i)^n + {\rm PMT}\, \frac{(1+i)^n - 1}{i} + {\rm FV} = 0, \tag{2}
$$
Относительно $i$ это уравнение в общем случае аналитические решено быть не может, поэтому надо использовать численные методы.

## Цель работы

 1. Написать функцию, решающее уравнение (2). Ваша функция должна вызывать `scipy.optimize.brentq`. Можете воспользоваться другим методом. Обоснуйте свой выбор. 
 2. Рассчитать годовые процентные ставки для примеров 1)-3). Годовая процентная ставка равна $i_{\rm year}=mi$, где $m$ - количество платежей в год (в нашем случае $m=12$).
 3. Сравнить процентные ставки. Какой кредитный план выбрали бы вы?

## Решение

In [1]:
import numpy as np
import scipy as sp
from scipy.optimize import brentq

Проинициализируем переменные, для основного финансового уравнения (1):

In [2]:
pv = 100000
fv = 0
pmt1 = -1500
n1 = 10 * 12
pmt2 = -1250
n2 = 15 * 12
pmt3 = -1160
n3 = 20 * 12

Напишем функции, решающее уравнение (2) для 3 вариантов кредита:

In [3]:
def fun1(i):
    return pv*(1+i)**n1 + pmt1*(((1+i)**n1-1)/i) + fv 
def fun2(i):
    return pv*(1+i)**n2 + pmt2*(((1+i)**n2-1)/i) + fv 
def fun3(i):
    return pv*(1+i)**n3 + pmt3*(((1+i)**n3-1)/i) + fv 

Для решения уравнения (1) в задании рекомендуется вызывать `scipy.optimize.brentq(f, a, b)`. Функция использует классический метод Брента для нахождения нуля функции f на интервале смены знака [a, b]. Обычно считается лучшей из процедур поиска корня.

In [4]:
sol1 = brentq(fun1, 0.01, 1.0)
sol1

0.010930603939868333

In [5]:
sol2 = brentq(fun2, 0.01, 1.0)
sol2

0.01063980717924384

In [6]:
sol3 = brentq(fun3, 0.01, 1.0)
sol3

0.010697698792848072

Напишем функцию решающующую, уравненение (1)

In [7]:
def solve_fun(pv, pmt, n, fv):
    def fun(i):
        return pv*(1+i)**n + pmt*((((1+i)**n)-1)/i) + fv
    return brentq(fun, 0.001, 1.0)

Посчитаем ежемесячный процент, для каждого варианта

In [8]:
i1 = solve_fun(pv, pmt1, n1, fv)
print(sol1)
print(i1)

0.010930603939868333
0.010930603939676818


In [9]:
i2= solve_fun(pv, pmt2, n2, fv)
print(sol2)
print(i2)

0.01063980717924384
0.01063980717901046


In [10]:
i3= solve_fun(pv, pmt3, n3, fv)
print(sol3)
print(i3)

0.010697698792848072
0.010697698792974464


Вычислим годовой процент

In [11]:
iy1 = 12*i1
iy2 = 12*i2
iy3 = 12*i3
print("Кредит на 10 лет, годовая ставка =", iy1*100,'%')
print("Кредит на 15 лет, годовая ставка =", iy2*100,'%')
print("Кредит на 20 лет, годовая ставка =", iy3*100,'%')

Кредит на 10 лет, годовая ставка = 13.11672472761218 %
Кредит на 15 лет, годовая ставка = 12.767768614812553 %
Кредит на 20 лет, годовая ставка = 12.837238551569358 %


## Вывод

Годовые ставки для кредита на сумму $\$100000$. составили:
* $13.12 \% $ для варианта 1
* $12.77 \% $ для варианта 2
* $12.84 \% $ для варианта 3

Конечно, брать кредит на 15 лет,не хотелось бы. Но из данных вариантов наиболее выгодным в плане годовой ставки является 2 вариант. Так что я бы выбрал именно его.